source/fab/tasks/c.py

```diff
@@ -0,0 +1,535 @@
+# (c) Crown copyright Met Office. All rights reserved.
+# For further details please refer to the file COPYRIGHT
+# which you should have received as part of this distribution
+"""
+C language handling classes.
+"""
+import subprocess
+import re
+import clang.cindex  # type: ignore
+from collections import deque
+from typing import \
+    List, \
+    Iterator, \
+    Pattern, \
+    Optional, \
+    Match, \
+    Sequence, \
+    Generator, \
+    Union
+from pathlib import Path
+
+from fab.database import \
+    StateDatabase, \
+    DatabaseDecorator, \
+    FileInfoDatabase, \
+    SqliteStateDatabase, \
+    WorkingStateException
+from fab.tasks import Task, TaskException
+from fab.artifact import \
+    Artifact, \
+    Raw, \
+    Modified, \
+    Analysed, \
+    Compiled, \
+    BinaryObject
+from fab.reader import \
+    TextReader, \
+    FileTextReader, \
+    TextReaderDecorator
+
+
+class CSymbolUnresolvedID(object):
+    def __init__(self, name: str):
+        self.name = name
+
+    def __eq__(self, other):
+        if not isinstance(other, CSymbolUnresolvedID):
+            message = "Cannot compare CSymbolUnresolvedID with " \
+                + other.__class__.__name__
+            raise TypeError(message)
+        return other.name == self.name
+
+
+class CSymbolID(CSymbolUnresolvedID):
+    def __init__(self, name: str, found_in: Path):
+        super().__init__(name)
+        self.found_in = found_in
+
+    def __hash__(self):
+        return hash(self.name) + hash(self.found_in)
+
+    def __eq__(self, other):
+        if not isinstance(other, CSymbolID):
+            message = "Cannot compare CSymbolID with " \
+                + other.__class__.__name__
+            raise TypeError(message)
+        return super().__eq__(other) and other.found_in == self.found_in
+
+
+class CInfo(object):
+    def __init__(self,
+                 symbol: CSymbolID,
+                 depends_on: Sequence[str] = ()):
+        self.symbol = symbol
+        self.depends_on = list(depends_on)
```

Comment:
Is this, in fact, a list of symbols? Should it be `list[SymbolID]'?

Reply:
Having tried to implement this - I think conceptually it seems like it should be, but in practice it maybe doesn't make sense - when we are adding dependencies we won't know the `found_in` part of the `CSymbolID`and it feels wrong to attach it with a dummy "empty" path or something